File: src/leetcodepython/list/remove_nth_node_from_end_list_19.py
# encoding='utf-8'

'''
/**
 * This is the solution of No. 21 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/merge-two-sorted-lists
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 将两个升序链表合并为一个新的升序链表并返回。新链表是通过拼接给定的两个链表的所有节点组成的。 
 * <p>
 * 示例：
 * <p>
 * 输入：1->2->4, 1->3->4
 * 输出：1->1->2->3->4->4
 * <p>
 * 来源：力扣（LeetCode）
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
'''
from list.list_node import ListNode


class Solution:
    def remove_nth_from_end(self, head: ListNode, n: int) -> ListNode:
        '''
            删除倒数第n个点
        Args:
            head: 头节点
            n: 倒数n个数
        Returns:
            删除后的结果
        '''
        if head == None or head.next == None:
            return None
        dummy = ListNode(-1)
        dummy.next = head
        node1 = dummy
        node2 = dummy
        for i in range(n):
            node2 = node2.next
        while node2.next:
            node1 = node1.next
            node2 = node2.next
        node1.next = node1.next.next
        return dummy.next


if __name__ == '__main__':
    list_node = ListNode()
    l2 = list_node.create_list_node_array([1, 3, 4, 5])
    solution = Solution()
    result = solution.remove_nth_from_end(l2, 2)
    print(result.val)
    assert result.val == 1

File: src/leetcodejava/list/RemoveNthNodeFromEndList19.java
package leetcodejava.list;

import org.junit.Assert;
import org.junit.Test;

/**
 * This is the solution of No.19 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/remove-nth-node-from-end-of-list
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。
 * <p>
 * 进阶：你能尝试使用一趟扫描实现吗？
 * <p>
 * 示例 1：
 * <p>
 * 输入：head = [1,2,3,4,5], n = 2
 * 输出：[1,2,3,5]
 * 示例 2：
 * <p>
 * 输入：head = [1], n = 1
 * 输出：[]
 * 示例 3：
 * <p>
 * 输入：head = [1,2], n = 1
 * 输出：[1]
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/remove-nth-node-from-end-of-list
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class RemoveNthNodeFromEndList19 {
    @Test
    public void decompressRunLengthEncodedTest() {
        Integer[] nums = {1, 2, 3, 4};
        ListNode node = ListNode.createListNode(nums);
        ListNode res = removeNthFromEnd(node, 2);
        Assert.assertEquals(res.val, 1);
    }

    /**
     * 移除倒数第n数字
     *
     * @param head 头结点
     * @param n    第n个数字
     * @return 结果链表
     */
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null || head.next == null) {
            return null;
        }
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode p1 = dummy;
        ListNode p2 = head;
        for (int i = 0; i < n; ++i) {
            p2 = p2.next;
        }
        while (p2 != null) {
            p1 = p1.next;
            p2 = p2.next;
        }
        // 删除元素
        p1.next = p1.next.next;
        // 将元素置为空
        p1.next = null;
        return dummy.next;
    }
}
